devfile/devworkspace-operator
# Create an initcontainer for persistent home setup

### What does this PR do?

### The problem

Today, the persistent home setup occurs in the entrypoint [[1]](https://github.com/devfile/developer-images/blob/3d74f3745dc75fe20ca283d3425ca5dbce849f7f/base/ubi8/Dockerfile#L123) of the tooling container (assuming it is based on the UBI). Notably, the entrypoint runs stow [[2]](https://github.com/devfile/developer-images/blob/3d74f3745dc75fe20ca283d3425ca5dbce849f7f/universal/ubi8/entrypoint.sh#L62) to make symbolic links for the home directory to be saved on the PVC.

If a poststart event depends on files or edits files located in the home directory, there is a race condition because the poststart event may start running before stow has finished running. Stow takes about 6 seconds to complete.

This race condition causes https://github.com/eclipse-che/che/issues/22914, because the checode poststart event runs before the entrypoint script completes.

### Proposed solution
To solve the race condition altogether, the entrypoint script must be complete before the post start events run. To do this, an initcontainer that runs the entrypoint script can be used.

#### What is the container image for the initcontainer?
To have a minimal effect on workspace startup times, the container image for the initcontainer should be the same image as the tooling container's image to avoid pulling an extra image. Assuming that the image is already pulled, I've observed no significant startup impact (see demo video in the next section) when running an extra container running the tooling image, even if the tooling image is large (quay.io/devfile/universal-developer-image:ubi8-latest has a compressed size of 2.9GB).

When thinking about it in the Docker perspective, running a container is only creating a writable layer on top of the already-pulled image layers, therefore if the image is already pulled the image size would not have a significant impact on container startup. ~~I'm guessing~~ this is what also happens in OpenShift/CRI-O, because adding the initcontainer did not increase workspace startup time significantly.

### What issues does this PR fix or reference?
https://github.com/eclipse-che/che/issues/22914
Fixes https://github.com/devfile/devworkspace-operator/issues/1260

Demo: https://youtu.be/RQeKKJgaogM

In the demo above, the init container named `init-persistent-home` runs for than 9 seconds at the 0:20 second mark. Stow is only run once (because of this if statement [[3]](https://github.com/devfile/developer-images/blob/3d74f3745dc75fe20ca283d3425ca5dbce849f7f/universal/ubi8/entrypoint.sh#L60)) to set up the new PVC.

So for subsequent workspace starts, the `init-persistent-home` init container takes less than a second to finish. Here is a second demo showing that:

https://youtu.be/hKnb1mgBSqA?si=yuJ3EKmj_dX0qHj0&t=20

### Is it tested? How?

To test this PR with Eclipse Che:

Install DWO on your cluster by adding this catalog source and installing DWO from OperatorHub:
```
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: devworkspace-operator-catalog
  namespace: openshift-marketplace
spec:
  sourceType: grpc
  image: quay.io/dkwon17/devworkspace-operator-index:init-container
  publisher: Red Hat
  displayName: DevWorkspace Operator Catalog
  updateStrategy:
    registryPoll:
      interval: 5m
```
After waiting a few minutes, the DWO from the new catalog should appear in OperatorHub. Install the one from `DevWorkspace Operator Catalog`.

![image](https://github.com/devfile/devworkspace-operator/assets/83611742/9ec65640-50b3-4be5-a2cf-6e9cc07093dd)

Install Eclipse Che with chectl:
```
chectl server:deploy -p openshift --skip-devworkspace-operator
```
Enabled persistent user home by adding the following in the CheCluster CR:
```
spec:
  devEnvironments:
    persistUserHome:
      enabled: true
```

Create a workspace with the following repositories and confirm that the `podman info` command runs without errors after workspace startup (just like in the demo video):
* https://github.com/che-samples/web-nodejs-sample
* empty sample workspace
* https://github.com/dkwon17/repo-with-poststart
* https://gist.githubusercontent.com/dkwon17/9ecac5397bef1178f77f308565814996/raw/4801b3c07a2cababd53b3594754c628cde7b8947/child-devfile.yaml


### PR Checklist


### Is it tested? How?


### PR Checklist
